### PR TITLE
fix: tokens list isReady state

### PR DESCRIPTION
**Pull request by ChefJerry**
### Summary
🐛🧪🚀

This pull request introduces `vitest` as a new testing framework for the project and refactors the list reducer tests to use it. It also fixes some bugs and improves performance in the token list fetching and updating logic. It affects the files `packages/token-lists/react/reducer.test.ts`, `pnpm-lock.yaml`, `apps/web/src/state/lists/updater.ts`, `packages/token-lists/package.json`, and `packages/token-lists/react/reducer.ts`.

> _To test the token-lists package_
> _We added vitest to the stack_
> _And fixed some dependencies_
> _With better efficiencies_
> _And checked the lists before we fetch_

### Walkthrough
*  Add a test script and a testing framework for the token-lists package using vitest ([link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32d135d93ad68d85298f8be14539d906f2a3f8ffca7bfca0761362e20bcb1792L11-R33), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR1408-R1410), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR11264-R11271), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR11281-R11288), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR11297-R11304), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR11311-R11316), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR11325-R11332), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL25184-R25287), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR25414-R25418), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR25429-R25433), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR25933-R25935), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR26466-R26486), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL26600-R26805))
*  Refactor the list reducer tests to use the configureStore function and the preloaded state ([link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL2-R7), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL17-R16), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL26-R25), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL47-R63), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL65-R76), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL74-R85), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beR214), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL211-R223), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL240-R271), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL254-R283), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL282-R311), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL310-R339), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL328-R357), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL349-R378), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL374-R403), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL405-R434), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL434-R463), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-bcc050ff6a68e8eaab72023d667efc99471f4ba727235fffb11ae3b81f1ec8beL484-R513))
*  Export the ListByUrlState type and the NEW_LIST_STATE constant from the list reducer ([link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-fedc48a5e69a5650a392b3d0e57a3b99e5560fb069a3fc630302de3178165862L30-R30))
*  Add a condition to the useSWRImmutable hook to avoid fetching the token list when the lists are empty ([link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-ea31fe29d993a0fb1d1fff742dde6356fa713d6997722023b3ebf9d64d315a2aL51-R51))
*  Update the commit hash of the @release-it/conventional-changelog dependency to fix a bug with the release notes generation ([link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL1960-R1963), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL27428-R27631))
*  Update the versions of the acorn, pathe, diff-sequences, pretty-format, jest-diff, and mlly dependencies to fix bugs and improve performance ([link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL12430-R12498), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR12529-R12533), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL15407-R15472), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL17612-R17676), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL18892-R18956), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL19029-R19093), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL19071-R19137), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL19109-R19173), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL19274-R19338), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL19294-R19358), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL19322-R19386), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL19615-R19679), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL19676-R19740), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL19870-R19934), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL19912-R19976), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR20737-R20743), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL21285-R21368), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL26272-R26389), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL26293-R26410), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL26314-R26431), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL26814-R27016))
*  Add the dev flag to the @jest/schemas and pathe dependencies to mark them as development dependencies ([link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR6222-R6228), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL22301-R22393))
*  Add the @types/chai, @sinclair/typebox, acorn, magic-string, std-env, tinypool, tinyspy, and ufo dependencies to lock their versions and integrities ([link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL9895-R9912), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR9919-R9922), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR6427-R6429), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR8084-R8087), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR22601-R22609), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbR24582-R24585), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL24699-R24802), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL22187-R22270), [link](https://github.com/pancakeswap/pancake-frontend/pull/7830/files?diff=unified&w=0#diff-32824c984905bb02bc7ffcef96a77addd1f1602cff71a11fbbfdd7f53ee026bbL25960-R26076))

### CODE REVIEW

**Comment by chefjackson** [comment]
Let's checkout new branches on the remote repo directly next time. We do not recommend using forked repo as it requires to authorize the deployment every time

File: apps/web/src/state/lists/updater.ts [hunk 1]
```diff
@@ -48,7 +48,7 @@ export default function Updater(): null {
   })
 
   useSWRImmutable(
-    includeListUpdater && isReady && listState !== initialState ? ['token-list'] : null,
+    includeListUpdater && isReady && listState !== initialState && Object.keys(lists).length ? ['token-list'] : null,
```
Comment (chefjackson) [inline apps/web/src/state/lists/updater.ts:51]: can we take a look at the reason why `isReady` is `true` but there's no list provided?

Comment (ChefJerry) [reply apps/web/src/state/lists/updater.ts:51]: `lists` may still empty Object, when `isReady` turns true

Comment (chefjackson) [reply apps/web/src/state/lists/updater.ts:51]: yeah but that's not intended I think. `isReady` should be `true` only when the `lists` are ready, no?

File: packages/token-lists/react/lists.ts [hunk 2]
```diff
@@ -21,60 +21,54 @@ export const createListsAtom = (storeName: string, reducer: any, initialState: a
...
-        db,
         getItem: async (key: string) => {
           const value = await db.getItem(key)
           if (value) {
             return value
           }
           return initialState
         },
-        setItem: (k: string, v: any) => {
+        setItem: async (k: string, v: any) => {
```
Comment (chefjackson) [inline packages/token-lists/react/lists.ts:38]: no need to change here maybe? As the `db.setItem` is returning a promise

Comment (ChefJerry) [reply packages/token-lists/react/lists.ts:38]: actually, it is a promise

https://github.com/localForage/localForage/blob/master/typings/localforage.d.ts#L20

Comment (ChefJerry) [reply packages/token-lists/react/lists.ts:38]: follow the `AsyncStorage` type defined, the return type of `setItem` is `Promise<void>`, so no need to return

https://github.com/pmndrs/jotai/blob/1d25284f170275186e0ff1715967c7e560debae6/src/vanilla/utils/atomWithStorage.ts#L17

Comment (ChefJerry) [reply packages/token-lists/react/lists.ts:38]: follow the `AsyncStorage` type defined, the return type of `setItem` is `Promise<void>`, so no need to return

https://github.com/pmndrs/jotai/blob/1d25284f170275186e0ff1715967c7e560debae6/src/vanilla/utils/atomWithStorage.ts#L17

**Review by chefjackson — APPROVED** [review]

**Pull request merged**